lib/Sema/ConstraintSystem.cpp

```diff
@@ -1110,7 +1112,8 @@ void ConstraintSystem::openGeneric(
        GenericSignature *sig,
        bool skipProtocolSelfConstraint,
        ConstraintLocatorBuilder locator,
-       OpenedTypeMap &replacements) {
+       OpenedTypeMap &replacements,
+       bool skipGenericRequirements) {
```

Comment:
Instead of adding a new boolean parameter here, how about removing the openGenericRequirements() call below, and then changing all callers to either call it or not? It would be clearer than yet another boolean. This allows you to remove the skipProtocolSelfConstraint parameter from openGeneric() as well, since its only used by openGenericRequirements().

Reply:
I have considered doing that, but I didn't go that route because it means that we'd have to do duplicate direct calls everywhere when it's only useful in this special case. I think it makes this whole setup more error-prone. Other places like CRanking/getTypeOfReference shouldn't have to do anything special besides a call to openFunctionType/openGeneric.

Reply:
But there aren't that many calls overall are there? I'm just not a fan of boolean flags as parameters to functions.

Reply:
You could also factor into three functions, openGeneric() that does both, openGenericParams() creating the type variables and binding the outer archetypes only, and openGenericRequirements()

Reply:
There aren't many calls at the moment, but I want to make sure that `openGeneric`/`openFunctionType` still does the right thing by default. I can absolutely look into refactoring this some more though.